mlir/lib/Conversion/MathToLibm/MathToLibm.cpp

```diff
@@ -185,8 +186,15 @@ void mlir::populateMathToLibmConversionPatterns(RewritePatternSet &patterns) {
   populatePatternsForOp<math::Log10Op>(patterns, ctx, "log10f", "log10");
   populatePatternsForOp<math::Log1pOp>(patterns, ctx, "log1pf", "log1p");
   populatePatternsForOp<math::PowFOp>(patterns, ctx, "powf", "pow");
-  populatePatternsForOp<math::RoundEvenOp>(patterns, ctx, "roundevenf",
-                                           "roundeven");
+  if (options.allowC23Features)
+    populatePatternsForOp<math::RoundEvenOp>(patterns, ctx, "roundevenf",
+                                             "roundeven");
+  else if (options.roundingModeIsDefault)
+    populatePatternsForOp<math::RoundEvenOp>(patterns, ctx, "nearbyintf",
+                                             "nearbyint");
+  // Roundeven: using nearbyint (pre-C23) for roundeven requires the
```

Comment:
I am not really qualified to review those changes, but since you asked for a review, perhaps it helps if I ask a few questions to understand what you are trying to do. I see that you need this for Flang. I suppose that roundeven provides the semantics that you need for Fortran. How is this related to "C23" features? Isn't this a question about what is implemented in your libm? It seems that glibc has implemented this (specifically for Fortran support) for several years, see [here](https://sourceware.org/legacy-ml/libc-alpha/2016-12/msg00791.html). Is it unimplemented on any target that you need?

Reply:
Thanks @frederik-h. I have touched `flang` code at some point in an attempt to fix a compile-time bug that went unnoticed on my end, but went backwards and found out I could fix it at the MathToLibm end. So this has nothing to do with Flang... sorry for the noise.

My use case is indeed a C compiler that does not support (yet) roundeven.

Reply:
Thanks for the explanation. So I assume that using `nearbyint` instead of `roundeven` should be equivalent as long as no one changes the rounding mode? Or are there other semantics differences? I am wondering if you could handle this in your frontend if there was a `nearbyint` in the math dialect? Not sure if that's better though.

Reply:
> My use case is indeed a C compiler that does not support (yet) roundeven. It comes with its own libm implementation that does not feature this function.

So this is a workaround for a missing implementation of `roundeven` in your libm? Some thoughts about alternative solutions: (1) Implement it in your libm, (2) keep this change in your projects fork of mlir until you implement it if that exists, or (3) add a `math.nearbyint` and document that its behavior depends on a target specific rounding mode (assuming that this would be at least as acceptable for the `math` dialect as the changes from this PR) and then replace `math.roundeven` by that in a pass in your front end? (3) sounds like the best approach to me assuming that (1) is not an option for you right now, but we would need the opinion of someone who is more familiar with the design goals of the dialect.

Reply:
Thanks for your insight. I unfortunately cannot do option (1); option (3) would result in having two very similar operations in math. Option (2) is feasible.

Currently, MathToLibm is tied to specific versions of libm that feature the `roundeven` function; to me, it would make sense to be able to specify which standard / version of libm is targeted so that only compatible calls are issued. There would need to be more use cases to this than just mine to warrant such a big change though.